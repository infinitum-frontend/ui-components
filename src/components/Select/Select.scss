.inf-select {
  &__items {
    border-radius: var(--inf-border-radius-small);
    border: 1px solid var(--inf-color-border-default);
    background-color: white;
    width: 100%;
    list-style-type: none;
    z-index: var(--inf-z-index-dropdown);

    // TODO: обсудить, почему снизу есть паддинг а сверху нет. Как выглядит ховер на последний элемент и на первый
    padding-bottom: 4px;

    &:focus-visible {
      outline: none;
    }

    &::-webkit-scrollbar {
      width: 7px;
      height: 7px;
      background: #e7e7e7;
      border-radius: 5px;
    }

    &::-webkit-scrollbar-track {
      border-radius: 5px;
    }

    &::-webkit-scrollbar-thumb {
      border-radius: 5px;
      width: 7px;
      height: 7px;
      background-color: #c9c9c9;
    }
  }

  &__item {
    cursor: pointer;
    font-size: var(--inf-text-size-small);
    padding: 7px 12px;
    line-height: 18px;
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow-x: hidden;

    &--active {
      background-color: var(--inf-control-background-color-hover);
    }

    &:first-child {
      border-top-right-radius: var(--inf-select-item-border-radius);
      border-top-left-radius: var(--inf-select-item-border-radius);
    }
  }

  select {
    opacity: 0;
    outline: none;
    position: absolute;
    left: 0;
    margin: 0;
    z-index: -1;
  }
}
