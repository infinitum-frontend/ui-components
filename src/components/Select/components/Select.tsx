// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, {
  KeyboardEventHandler,
  forwardRef,
  useEffect,
  useState,
  ReactElement,
  useRef,
  FocusEventHandler,
  MouseEventHandler,
  FormEventHandler
} from 'react'
import '../style/index.scss'
import cn from 'classnames'
import { SelectProps, SelectOption } from '../interface'
import { mergeRefs } from 'react-merge-refs'
import { useFormGroup } from 'Components/Form/context/group'
import {
  autoUpdate,
  flip,
  offset,
  useFloating,
  FloatingPortal,
  useInteractions,
  useDismiss,
  size
} from '@floating-ui/react'
import SelectButton from 'Components/Select/components/SelectButton'

export const defaultSelectItem: SelectOption = {
  value: -1,
  label: 'Не указано'
}

const getIndexByValue = (
  value: SelectOption['value'] | undefined,
  items: SelectOption[]
): number => {
  return items.findIndex((item) => item.value === value)
}

const getItemByValue = (
  value: SelectOption['value'],
  items: SelectOption[]
): SelectOption | undefined => {
  return items.find((item) => item.value === value)
}

/** Компонент для выбора значения из выпадающего списка */
const Select = forwardRef<HTMLButtonElement, SelectProps>(
  (
    {
      options = [],
      className = '',
      onChange,
      autoFocus = false,
      value,
      disabled = false,
      placeholder = defaultSelectItem.label as string,
      required = false,
      status,
      maxItemsCount = 12,
      ...props
    }: SelectProps,
    ref
  ): ReactElement => {
    // ============================= state =============================
    const [isOpened, setOpened] = useState(false)
    const [isFocused, setFocused] = useState(false)
    const [activeItem, setActiveItem] = useState(
      value ? getIndexByValue(value, options) : 0
    )

    const formGroupData = useFormGroup()
    const displayRef = useRef<HTMLButtonElement>(null)
    const selectRef = useRef<HTMLSelectElement>(null)

    // ============================= effects =============================
    useEffect(() => {
      if (autoFocus) {
        setOpened(true)
        setFocused(true)
      }
    }, [autoFocus])

    // ============================= handlers =============================
    const handleClick: MouseEventHandler<HTMLButtonElement> = (e): void => {
      if (
        displayRef.current?.contains(e.target as HTMLElement) &&
        (e.target as HTMLElement).tagName !== 'SELECT'
      ) {
        setOpened((prev) => !prev)
      }
    }

    // фокус, который поднимается от внутреннего нативного селекта
    const handleFocus: FocusEventHandler = (e) => {
      e.preventDefault()
      setFocused(true)
    }

    // блюр, который поднимается от внутреннего нативного селекта. Если было нажатие на элементы выпадающего списка, фокус не скидывается
    const handleBlur: FocusEventHandler = (e) => {
      const classList = e.relatedTarget?.classList

      if (classList?.contains('inf-select__items')) {
        return
      }

      setFocused(false)
    }

    const handleKeyDown: KeyboardEventHandler = (e) => {
      switch (e.key) {
        case 'ArrowDown':
          e.preventDefault()
          setActiveItem((prevState) => {
            if (prevState === options.length - 1) {
              return 0
            }

            return prevState + 1
          })
          break
        case 'ArrowUp':
          e.preventDefault()
          setActiveItem((prevState) => {
            if (prevState === 0) {
              return options.length - 1
            }

            return prevState - 1
          })
          break
        case 'Escape':
          e.preventDefault()
          setActiveItem(0)
          setOpened(false)
          break
        case ' ':
          e.preventDefault()
          setOpened(true)
          break
        case 'Enter':
          e.preventDefault()
          if (isOpened) {
            submit(activeItem)
          } else {
            setOpened(true)
          }
          break
        case 'Tab':
          if (isOpened) {
            setOpened(false)
          }
      }
    }

    const handleItemSelect = (id: number | string): void => {
      const index = getIndexByValue(id, options)
      setActiveItem(index)
      // сохраняем состояния фокуса на элементе триггере при опции с помощью мышки
      displayRef.current?.focus()
      submit(index)
    }

    const handleItemMouseOver = (index: number): void => {
      setActiveItem(index)
    }

    const handleInvalid: FormEventHandler<HTMLSelectElement> = () => {
      if (formGroupData) {
        formGroupData.setInvalid?.(true)
        selectRef.current?.setCustomValidity(formGroupData.invalidMessage || '')
      }
    }

    const submit = (index: number): void => {
      if (index < 0) {
        return
      }

      setOpened(false)
      if (formGroupData) {
        formGroupData.setInvalid?.(false)
        selectRef.current?.setCustomValidity('')
      }
      onChange?.(options[index])
    }

    // ============================= floating =============================
    const { x, y, refs, context } = useFloating({
      open: isOpened,
      onOpenChange: setOpened,
      placement: 'bottom-start',
      whileElementsMounted: autoUpdate,
      middleware: [
        offset(4),
        flip(),
        size({
          apply({ rects, elements }) {
            Object.assign(elements.floating.style, {
              width: `${rects.reference.width}px`
            })
          }
        })
      ]
    })

    const { getReferenceProps, getFloatingProps } = useInteractions([
      useDismiss(context, {
        outsidePress: (e) => {
          if (formGroupData) {
            return (e.target as HTMLLabelElement)?.htmlFor !== formGroupData?.id
          } else {
            return true
          }
        }
      })
    ])

    // ============================= render =============================
    const isValueExists = Boolean(value) || Number.isInteger(value)
    // высота элемента, паддинг и границы
    const maxHeight = maxItemsCount * 32 + 4 + 2

    return (
      <>
        <SelectButton
          tabIndex={-1}
          ref={mergeRefs([ref, refs.setReference, displayRef])}
          disabled={disabled}
          selected={isValueExists}
          focused={isFocused}
          status={status}
          onFocus={handleFocus}
          onClick={handleClick}
          onBlur={handleBlur}
          className={cn(className, 'inf-select')}
          {...getReferenceProps({
            onKeyDown: handleKeyDown
          })}
          {...props}
        >
          {isValueExists
            ? getItemByValue(value as string | number, options)?.label
            : placeholder}
          <select
            ref={selectRef}
            required={formGroupData?.required || required}
            aria-required={formGroupData?.required || required}
            aria-invalid={formGroupData?.invalid || undefined}
            disabled={disabled}
            id={formGroupData?.id}
            value={value}
            onInvalid={handleInvalid}
            onChange={() => null}
          >
            <option value={''} />
            {options.map((option) => (
              <option value={option.value} key={option.value}>
                {option.label}
              </option>
            ))}
          </select>
        </SelectButton>

        <FloatingPortal>
          {isOpened && !disabled && (
            <ul
              style={{
                position: 'absolute',
                maxHeight: `${maxHeight}px`,
                overflowX: 'hidden',
                overflowY: 'auto',
                top: y ?? 0,
                left: x ?? 0
              }}
              ref={refs.setFloating}
              className={cn('inf-select__items')}
              {...getFloatingProps()}
            >
              {Boolean(options.length) &&
                options.map((option, index) => (
                  <li
                    key={option.value}
                    onMouseOver={() => handleItemMouseOver(index)}
                    onClick={() => handleItemSelect(option.value)}
                    className={cn('inf-select__item', {
                      'inf-select__item--active': index === activeItem
                    })}
                  >
                    {option.label}
                  </li>
                ))}
            </ul>
          )}
        </FloatingPortal>
      </>
    )
  }
)

Select.displayName = 'Select'

export default Select
