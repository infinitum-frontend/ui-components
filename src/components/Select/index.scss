.inf-select {
  position: relative;
  width: 100%;
  cursor: pointer;
  user-select: none;
  appearance: none;
  text-align: left;
  background-color: var(--inf-color-background-base);
  border-radius: var(--inf-border-radius-medium);
  border: 1px solid var(--inf-color-border-default);
  padding: 7px var(--inf-input-padding-horizontal-medium);
  line-height: 18px;
  height: var(--inf-control-height-medium);
  font-size: var(--inf-text-size-small);
  font-family: var(--inf-font-family-base);

  &:focus-visible {
    outline: none;
  }

  &--selected {
    border-color: var(--inf-control-color-border-active);
  }

  &--disabled {
    border-color: transparent;
    background-color: var(--inf-color-background-control-disabled);
    color: var(--inf-color-text-tertiary);
    pointer-events: none;
  }

  &--status-error {
    border-color: var(--inf-color-primary);
  }

  &__arrow {
    position: absolute;
    display: flex;
    align-items: center;
    justify-content: center;
    width: 16px;
    height: 16px;
    top: 50%;
    transform: translateY(-50%);
    right: 10px;

    &--selected {
      color: var(--inf-control-color-border-active);
    }
  }

  &__items {
    border-radius: var(--inf-border-radius-medium);
    border: 1px solid var(--inf-color-border-default);
    background-color: white;
    width: 100%;
    list-style-type: none;

    // TODO: обсудить, почему снизу есть паддинг а сверху нет. Как выглядит ховер на последний элемент и на первый
    padding-bottom: 4px;

    &:focus-visible {
      outline: none;
    }
  }

  &__item {
    display: flex;
    cursor: pointer;
    font-size: var(--inf-text-size-small);
    padding: 7px 12px;
    line-height: 18px;

    &--active {
      background-color: var(--inf-control-background-color-hover);
    }

    &:first-child {
      border-top-right-radius: var(--inf-select-item-border-radius);
      border-top-left-radius: var(--inf-select-item-border-radius);
    }
  }

  select {
    opacity: 0;
    outline: none;
    position: absolute;
    margin: 0;
    z-index: -1;
  }
}
