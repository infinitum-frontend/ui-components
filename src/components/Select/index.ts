export { default as InfSelect } from './InfSelect'
