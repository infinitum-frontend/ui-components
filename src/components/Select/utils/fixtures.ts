// export const SelectBaseOptions = [
//   { label: 'Депозитарные услуги', value: 'one' },
//   { label: 'Очень длинные спецдепозитарные услуги', value: 'two' },
//   { label: 'Консалтинг и аутсорсинг', value: 'three' }
// ]

export const SelectBaseOptions = [
  { label: 'Депозитарные услуги', value: 0 },
  { label: 'Спецдепозитарные услуги', value: 1 },
  { label: 'Консалтинг и аутсорсинг', value: 2 }
]

export const SelectBaseGroupedOptions = [
  { label: 'Депозитарные услуги', value: 'one' },
  {
    label: 'Спецдепозитарные услуги',
    options: [
      { label: 'Спецдепозитарные услуги 1', value: 'two-one' },
      { label: 'Очень длинные спецдепозитарные услуги 2', value: 'two-two' },
      { label: 'Спецдепозитарные услуги 3', value: 'two-three' }
    ]
  },
  {
    label: 'Консалтинг и аутсорсинг',
    options: [{ label: 'Консалтинг и аутсорсинг 1', value: 'three-one' }]
  }
]

export const SelectLongOptions = [
  { label: 'Депозитарные услуги', value: 'one' },
  { label: 'Очень длинные спецдепозитарные услуги', value: 'two' },
  { label: 'Консалтинг и аутсорсинг', value: 'three' },
  { label: 'Консалтинг и аутсорсинг 2', value: 'four' },
  { label: 'Консалтинг и аутсорсинг 3', value: 'five' },
  { label: 'Консалтинг и аутсорсинг 4', value: 'six' },
  { label: 'Консалтинг и аутсорсинг 5', value: 'sevent' },
  { label: 'Консалтинг и аутсорсинг 6', value: 'eight' },
  { label: 'Консалтинг и аутсорсинг 7', value: 'nine' },
  { label: 'Консалтинг и аутсорсинг 8', value: 'ten' }
]

export const SelectOptionsUnformatted = [
  { text: 'Депозитарные услуги', id: 0 },
  { text: 'Спецдепозитарные услуги', id: 1 },
  { text: 'Консалтинг и аутсорсинг', id: 2 }
]

export const SelectOptionsWithExtraKeys = [
  { text: 'Депозитарные услуги', id: 0, subtext: 'Subtext' },
  { text: 'Спецдепозитарные услуги', id: 1, subtext: 'Subtext' },
  { text: 'Консалтинг и аутсорсинг', id: 2, subtext: 'Subtext' }
]

export const SelectOptionsRawLong = [
  {
    id: '1',
    name: 'Ошибка доставки'
  },
  {
    id: '2',
    name: 'На отправке'
  },
  {
    id: '3',
    name: 'Доставлен'
  },
  {
    id: '4',
    name: 'Получен'
  },
  {
    id: '5',
    name: 'Ошибка в обработке'
  },
  {
    id: '6',
    name: 'Отправлен'
  },
  {
    id: '7',
    name: 'Отказано в обработке'
  },
  {
    id: '8',
    name: 'Сохранено'
  }
]
