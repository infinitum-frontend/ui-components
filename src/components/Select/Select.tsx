// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, {
  KeyboardEventHandler,
  forwardRef,
  useEffect,
  useState,
  ReactElement,
  useRef,
  FocusEventHandler,
  MouseEventHandler,
  useContext,
  useId
} from 'react'
import './Select.scss'
import cn from 'classnames'
import { SelectProps, SelectOption } from './types'
import { mergeRefs } from 'react-merge-refs'
import {
  autoUpdate,
  flip,
  offset,
  useFloating,
  FloatingPortal,
  useInteractions,
  useDismiss,
  size
} from '@floating-ui/react'
import SelectButton from './components/SelectButton'
import FormGroupContext from 'Components/Form/context/group'
import FormContext from 'Components/Form/context/form'
import useFormControlHandlers from 'Components/Form/hooks/useFormControlHandlers'

export const defaultSelectItem: SelectOption = {
  value: -1,
  label: 'Не указано'
}

const getIndexByValue = (
  value: SelectOption['value'] | undefined,
  items: SelectOption[]
): number => {
  return items.findIndex((item) => item.value === value)
}

const getItemByValue = (
  value: SelectOption['value'],
  items: SelectOption[]
): SelectOption | undefined => {
  return items.find((item) => item.value === value)
}

/** Компонент для выбора значения из выпадающего списка */
const Select = forwardRef<HTMLButtonElement, SelectProps>(
  (
    {
      options = [],
      className = '',
      onChange,
      autoFocus = false,
      value,
      disabled: disabledProp = false,
      loading = false,
      placeholder = defaultSelectItem.label as string,
      required = false,
      status,
      maxItemsCount = 12,
      'aria-required': ariaRequired,
      'aria-invalid': ariaInvalid,
      ...props
    }: SelectProps,
    ref
  ): ReactElement => {
    // ============================= state =============================
    const [isOpened, setOpened] = useState(false)
    const [isFocused, setFocused] = useState(false)
    const [activeItem, setActiveItem] = useState(
      value ? getIndexByValue(value, options) : 0
    )

    const prefix = useId()

    const formGroupData = useContext(FormGroupContext)
    const formData = useContext(FormContext)
    const { onControlInvalid, resetControlValidity } = useFormControlHandlers()
    const disabled = disabledProp || formData?.disabled

    const displayRef = useRef<HTMLButtonElement>(null)
    const selectRef = useRef<HTMLSelectElement>(null)

    // ============================= effects =============================
    useEffect(() => {
      if (autoFocus) {
        setOpened(true)
        setFocused(true)
      }
    }, [autoFocus])

    // ============================= handlers =============================
    const handleClick: MouseEventHandler<HTMLButtonElement> = (e): void => {
      if (
        displayRef.current?.contains(e.target as HTMLElement) &&
        (e.target as HTMLElement).tagName !== 'SELECT'
      ) {
        setOpened((prev) => !prev)
      }
    }

    // фокус, который поднимается от внутреннего нативного селекта
    const handleFocus: FocusEventHandler = (e) => {
      e.preventDefault()
      setFocused(true)
    }

    // блюр, который поднимается от внутреннего нативного селекта. Если было нажатие на элементы выпадающего списка, фокус не скидывается
    const handleBlur: FocusEventHandler = (e) => {
      const target = e.relatedTarget
      if (target?.getAttribute?.('data-selector') === 'inf-select-items') {
        return
      }

      setFocused(false)
    }

    const handleKeyDown: KeyboardEventHandler = (e) => {
      switch (e.key) {
        case 'ArrowDown':
          e.preventDefault()
          setActiveItem((prevState) => {
            if (prevState === options.length - 1) {
              return 0
            }

            return prevState + 1
          })
          break
        case 'ArrowUp':
          e.preventDefault()
          setActiveItem((prevState) => {
            if (prevState === 0) {
              return options.length - 1
            }

            return prevState - 1
          })
          break
        case 'Escape':
          e.preventDefault()
          setActiveItem(0)
          setOpened(false)
          break
        case ' ':
          e.preventDefault()
          setOpened(true)
          break
        case 'Enter':
          e.preventDefault()
          if (isOpened) {
            submit(activeItem)
          } else {
            setOpened(true)
          }
          break
        case 'Tab':
          if (isOpened) {
            setOpened(false)
          }
      }
    }

    const handleItemSelect = (id: number | string): void => {
      const index = getIndexByValue(id, options)
      setActiveItem(index)
      // сохраняем состояния фокуса на элементе триггере при опции с помощью мышки
      displayRef.current?.focus()
      submit(index)
    }

    const handleItemMouseOver = (index: number): void => {
      setActiveItem(index)
    }

    const submit = (index: number): void => {
      if (index < 0) {
        return
      }

      setOpened(false)
      resetControlValidity()
      onChange?.(options[index])
    }

    // ============================= floating =============================
    const { x, y, refs, context } = useFloating({
      open: isOpened,
      onOpenChange: setOpened,
      placement: 'bottom-start',
      whileElementsMounted: autoUpdate,
      middleware: [
        offset(4),
        flip(),
        size({
          apply({ rects, elements }) {
            Object.assign(elements.floating.style, {
              width: `${rects.reference.width}px`
            })
          }
        })
      ]
    })

    const { getReferenceProps, getFloatingProps } = useInteractions([
      useDismiss(context, {
        outsidePress: (e) => {
          if (formGroupData) {
            return (e.target as HTMLLabelElement)?.htmlFor !== formGroupData?.id
          } else {
            return true
          }
        }
      })
    ])

    // ============================= render =============================
    const isValueExists = Boolean(value) || Number.isInteger(value)
    const displayValue = isValueExists
      ? getItemByValue(value as string | number, options)?.label
      : placeholder
    // высота элемента, паддинг и границы
    const maxHeight = maxItemsCount * 32 + 4 + 2

    return (
      <>
        <SelectButton
          tabIndex={-1}
          ref={mergeRefs([ref, refs.setReference, displayRef])}
          disabled={disabled}
          loading={loading}
          selected={isValueExists}
          focused={isFocused}
          status={status}
          onFocus={handleFocus}
          onBlur={handleBlur}
          title={typeof displayValue === 'string' ? displayValue : ''}
          className={cn(className, 'inf-select')}
          {...getReferenceProps({
            onKeyDown: handleKeyDown,
            onClick(e) {
              e.stopPropagation()
              handleClick(e as any)
            }
          })}
          {...props}
        >
          {displayValue}
          <select
            ref={selectRef}
            required={formGroupData?.required || required}
            aria-required={formGroupData?.required || required || ariaRequired}
            aria-invalid={formGroupData?.invalid || ariaInvalid}
            disabled={disabled}
            id={formGroupData?.id}
            value={value}
            onInvalid={onControlInvalid}
            onChange={() => {}}
          >
            <option value={''} />
            {options.map((option) => (
              <option value={option.value} key={prefix + String(option.value)}>
                {option.label}
              </option>
            ))}
          </select>
        </SelectButton>

        <FloatingPortal>
          {isOpened && !disabled && (
            <ul
              style={{
                position: 'absolute',
                maxHeight: `${maxHeight}px`,
                overflowX: 'hidden',
                overflowY: 'auto',
                top: y ?? 0,
                left: x ?? 0
              }}
              ref={refs.setFloating}
              className="inf-select__items"
              data-selector="inf-select-items"
              {...getFloatingProps({
                onClick(e) {
                  e.stopPropagation()
                }
              })}
            >
              {Boolean(options.length) &&
                options.map((option, index) => (
                  <li
                    key={String(option.value) + prefix}
                    title={typeof option.label === 'string' ? option.label : ''}
                    onMouseOver={() => handleItemMouseOver(index)}
                    onClick={() => handleItemSelect(option.value)}
                    className={cn('inf-select__item', {
                      'inf-select__item--active': index === activeItem
                    })}
                  >
                    {option.label}
                  </li>
                ))}
            </ul>
          )}
        </FloatingPortal>
      </>
    )
  }
)

Select.displayName = 'Select'

export default Select
