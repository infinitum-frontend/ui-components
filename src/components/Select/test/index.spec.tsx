import { describe, expect, it, vi } from 'vitest'
import { act, screen } from '@testing-library/react'
import { Select } from '../index'
import { SelectBaseOptions } from '../fixtures'
import { renderComponent } from '../../../../testSetup'
import { defaultSelectItem } from '../Select'
import userEvent from '@testing-library/user-event'
import { Form } from '../../Form'
import { Button } from '../../Button'
import { TextFieldClasses } from '../../../utils/textFieldClasses'

const title = 'select'
const user = userEvent.setup()

describe('select', () => {
  it('should match snapshot', () => {
    renderComponent(<Select options={SelectBaseOptions} title={title} />)

    expect(screen.queryByTitle(title)).toMatchSnapshot()
  })

  it('should render plain button by default', () => {
    renderComponent(<Select options={SelectBaseOptions} title={title} />)

    expect(screen.queryByTitle(title)).toBeInTheDocument()
    expect(screen.queryAllByRole('option')).toHaveLength(4)
  })

  it('should display default value if no selected', () => {
    renderComponent(<Select options={SelectBaseOptions} title={title} />)

    expect(
      screen.queryByText(defaultSelectItem.label as string)
    ).toBeInTheDocument()
  })

  it('should support className', () => {
    const customClassName = 'customClassName'
    renderComponent(
      <Select
        options={SelectBaseOptions}
        title={title}
        className={customClassName}
      />
    )

    expect(screen.queryByTitle(title)).toHaveClass(customClassName)
  })

  it('should support autoFocus', () => {
    renderComponent(
      <Select options={SelectBaseOptions} title={title} autoFocus={true} />
    )

    const wrapper = screen.queryByTitle(title)
    expect(wrapper).toHaveClass(TextFieldClasses.focused)

    expect(screen.queryAllByRole('listitem')[0]).toBeInTheDocument()
  })

  it('should support value', async () => {
    renderComponent(
      <Select options={SelectBaseOptions} value={1} title={title} />
    )

    const options = screen.queryAllByText(SelectBaseOptions[1].label)
    const select = screen.queryByRole('combobox') as HTMLSelectElement
    expect(options).toHaveLength(2)
    expect(
      options.find((option) => option.tagName === 'BUTTON')
    ).toBeInTheDocument()
    expect(select.value).toBe('1')

    const button = screen.queryByTitle(title) as HTMLButtonElement
    await user.click(button)

    const activeItem = screen.queryAllByRole('listitem')[1]
    expect(activeItem).toHaveClass('inf-select__item--active')
  })

  it('should support disabled', () => {
    renderComponent(
      <Select options={SelectBaseOptions} disabled={true} title={title} />
    )

    const button = screen.queryByTitle(title)
    expect(button).toBeDisabled()
    expect(button).toHaveClass(TextFieldClasses.disabled)
    expect(screen.queryByRole('combobox')).toBeDisabled()
  })

  it('should not display items when disabled', () => {
    renderComponent(
      <Select
        options={SelectBaseOptions}
        disabled={true}
        title={title}
        autoFocus={true}
      />
    )
    expect(screen.queryByRole('list')).not.toBeInTheDocument()
  })

  it('should support placeholder', () => {
    renderComponent(
      <Select options={SelectBaseOptions} placeholder={'Placeholder'} />
    )

    expect(screen.queryByText('Placeholder')).toBeInTheDocument()
  })

  it('should support required', () => {
    renderComponent(<Select options={SelectBaseOptions} required={true} />)
    const select = screen.queryByRole('combobox')

    expect(select).toHaveAttribute('required')
    expect(select).toHaveAttribute('aria-required', 'true')
  })

  it('should support status', () => {
    renderComponent(
      <Select options={SelectBaseOptions} status={'error'} title={title} />
    )
    expect(screen.queryByTitle(title)).toHaveClass(
      TextFieldClasses.status.error
    )
  })
})

describe('Select actions', () => {
  it('should display items on button click', async () => {
    renderComponent(<Select options={SelectBaseOptions} title={title} />)
    await user.click(screen.queryByTitle(title) as HTMLButtonElement)

    const list = screen.queryByRole('list')
    expect(list).toBeInTheDocument()
  })

  it('should support focus', () => {
    renderComponent(<Select options={SelectBaseOptions} />)
    act(() => {
      screen.queryByRole('combobox')?.focus()
    })
    expect(screen.queryByRole('button')).toHaveClass(TextFieldClasses.focused)
  })

  it('should support blur', () => {
    renderComponent(<Select options={SelectBaseOptions} />)
    act(() => {
      screen.queryByRole('combobox')?.focus()
    })
    act(() => {
      screen.queryByRole('combobox')?.blur()
    })

    expect(screen.queryByRole('button')).not.toHaveClass(
      'inf-select-button--focused'
    )
  })

  it('should close dropdown and remove focus when click outside', async () => {
    renderComponent(
      <div>
        <Select options={SelectBaseOptions} />
        <a href={''}>link</a>
      </div>
    )
    const button = screen.queryByRole('button') as HTMLButtonElement
    await user.click(button)
    expect(button).toHaveClass(TextFieldClasses.focused)
    expect(screen.queryByRole('list')).toBeInTheDocument()
    await user.click(screen.queryByRole('link') as HTMLElement)

    expect(button).not.toHaveClass(TextFieldClasses.focused)
    expect(screen.queryByRole('list')).not.toBeInTheDocument()
  })

  it('should support ArrowDown press', async () => {
    renderComponent(<Select options={SelectBaseOptions} />)
    const button = screen.queryByRole('button') as HTMLButtonElement
    await user.click(button)
    await user.keyboard('{ArrowDown}')

    expect(screen.queryAllByRole('listitem')[1]).toHaveClass(
      'inf-select__item--active'
    )

    await user.keyboard('{ArrowDown}')
    await user.keyboard('{ArrowDown}')
    expect(screen.queryAllByRole('listitem')[0]).toHaveClass(
      'inf-select__item--active'
    )
  })

  it('should support ArrowUp press', async () => {
    renderComponent(<Select options={SelectBaseOptions} />)
    const button = screen.queryByRole('button') as HTMLButtonElement
    await user.click(button)
    await user.keyboard('{ArrowUp}')

    expect(screen.queryAllByRole('listitem')[2]).toHaveClass(
      'inf-select__item--active'
    )

    await user.keyboard('{ArrowUp}')
    expect(screen.queryAllByRole('listitem')[1]).toHaveClass(
      'inf-select__item--active'
    )
  })

  it('should support Esc press', async () => {
    renderComponent(<Select options={SelectBaseOptions} />)
    const button = screen.queryByRole('button') as HTMLButtonElement
    await user.click(button)
    await user.keyboard('{Escape}')

    expect(screen.queryByRole('list')).not.toBeInTheDocument()
  })

  it('should support Space press', async () => {
    renderComponent(<Select options={SelectBaseOptions} />)
    const button = screen.queryByRole('button') as HTMLButtonElement
    act(() => {
      button.focus()
    })
    await user.keyboard(' ')

    expect(screen.queryByRole('list')).toBeInTheDocument()
  })

  it('should support Enter press', async () => {
    const onChange = vi.fn()
    renderComponent(<Select options={SelectBaseOptions} onChange={onChange} />)
    const button = screen.queryByRole('button') as HTMLButtonElement
    act(() => {
      button.focus()
    })

    await user.keyboard('{Enter}')
    expect(screen.queryByRole('list')).toBeInTheDocument()

    await user.keyboard('{Enter}')
    expect(screen.queryByRole('list')).not.toBeInTheDocument()
    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenCalledWith(SelectBaseOptions[0])
  })

  it('should support Tab press', async () => {
    const user = userEvent.setup()
    renderComponent(<Select options={SelectBaseOptions} />)
    const button = screen.queryByRole('button') as HTMLButtonElement
    await user.tab()
    expect(button).toHaveClass(TextFieldClasses.focused)
    await user.tab()
    expect(button).not.toHaveClass(TextFieldClasses.focused)

    await user.click(button)
    await user.tab()
    expect(button).toHaveClass(TextFieldClasses.focused)
    expect(screen.queryByRole('list')).not.toBeInTheDocument()
    await user.tab()
    expect(button).not.toHaveClass(TextFieldClasses.focused)
  })
})

describe('Select Items', () => {
  it('should select proper item on click', async () => {
    const onChange = vi.fn()
    renderComponent(<Select options={SelectBaseOptions} onChange={onChange} />)

    await user.click(screen.queryByRole('button') as HTMLButtonElement)

    await user.click(screen.queryAllByRole('listitem')[1])
    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenCalledWith(SelectBaseOptions[1])
  })

  it('should preserve focus on item click', async () => {
    renderComponent(<Select options={SelectBaseOptions} />)
    const button = screen.queryByRole('button') as HTMLButtonElement
    await user.click(button)
    await user.click(screen.queryAllByRole('listitem')[0])

    expect(button).toHaveClass(TextFieldClasses.focused)
    expect(document.activeElement).toHaveClass('inf-select')
  })

  it('should set active on Mouseover', async () => {
    renderComponent(<Select options={SelectBaseOptions} />)
    await user.click(screen.queryByRole('button') as HTMLButtonElement)

    const item = screen.queryAllByRole('listitem')[1]
    await user.hover(item)
    expect(item).toHaveClass('inf-select__item--active')
  })
})

describe('Select in Form', () => {
  it('should apply attrs if FormGroup is required', async () => {
    const { rerender } = renderComponent(
      <Form>
        <Form.Group required customValidationMessage={'Error'}>
          <Select options={SelectBaseOptions} title={title} />
        </Form.Group>

        <Button title={'submit'} type={'submit'} />
      </Form>
    )

    const select = screen.queryByRole('combobox') as HTMLSelectElement
    expect(select).toHaveAttribute('required')
    expect(select).toHaveAttribute('aria-required', 'true')
    expect(select).not.toHaveAttribute('aria-invalid')
    expect(select).toHaveAttribute('id')

    await user.click(screen.queryByTitle('submit') as HTMLButtonElement)
    expect(select).toHaveAttribute('aria-invalid', 'true')
    expect(select.validationMessage).toBe('Error')

    await user.click(screen.queryByTitle(title) as HTMLButtonElement)
    await user.click(screen.queryAllByRole('listitem')[0])

    rerender(
      <Form>
        <Form.Group required>
          <Select options={SelectBaseOptions} title={title} value={0} />
        </Form.Group>

        <Button title={'submit'} type={'submit'} />
      </Form>
    )

    expect(select).not.toHaveAttribute('aria-invalid')
    expect(select.validationMessage).toBe('')
  })

  it('should not close on label click', async () => {
    renderComponent(
      <Form>
        <Form.Group required customValidationMessage={'Error'}>
          <Form.Label role={'label'}>Label</Form.Label>
          <Select options={SelectBaseOptions} title={title} />
        </Form.Group>

        <Button title={'submit'} type={'submit'} />
      </Form>
    )

    await user.click(screen.queryByTitle(title) as HTMLButtonElement)
    await user.click(screen.queryByRole('label') as HTMLLabelElement)

    expect(screen.queryByRole('list')).toBeInTheDocument()
  })
})
