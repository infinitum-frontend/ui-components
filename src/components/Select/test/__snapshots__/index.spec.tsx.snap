// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`select > should match snapshot 1`] = `
<button
  class="inf-select-button inf-text-field inf-text-field--br-regular inf-select"
  tabindex="-1"
  title="select"
  type="button"
>
  <span
    class="inf-select-button__arrow"
  >
    <svg
      fill="none"
      height="5px"
      viewBox="0 0 12 7"
      width="10px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1 1L6 6L11 1"
        stroke="currentColor"
        stroke-linejoin="round"
        stroke-width="1.5"
      />
    </svg>
  </span>
  Не указано
  <select
    aria-required="false"
  >
    <option
      value=""
    />
    <option
      value="0"
    >
      Депозитарные услуги
    </option>
    <option
      value="1"
    >
      Спецдепозитарные услуги
    </option>
    <option
      value="2"
    >
      Консалтинг и аутсорсинг
    </option>
  </select>
</button>
`;
